### ISSUE TITLE
Support Result Set Management for PubSub item Queries

### ISSUE THREAD

**Issue by sebastianriese** [horazont/aioxmpp#255]
XEP-0084 specifies, that a service may return a subset of all elements when queried to retrieve all items (and page the results with result set management). 

As a first remedy `PubSubClient.get_items` should accept an RSM argument to be tacked to the request (and the XSO has to be extended to supply one). As a long term remedy `PubSubClient` should provide a method returning an async iterator (compare #11) that uses RSM to retrieve the items (or blocks of items) as they are requested.

(Crazy idea: Can we build a general RSM async iterator, that is given a query XSO with a magic RSM descriptor, and then resends the request as required by RSM to page through the large result set.)

**Comment by horazont** [comment]
I gave RSM some thought already. I think I am close to what you call a crazy idea :-).

I think the best way to deal with this is to create a generic RSMIterator class which takes a single argument, which is a coroutine.

This coroutine accepts another single argument, which is an RSM XSO, but this argument may be None. The coroutine must return a tuple consisting of an iterable and an RSM XSO as returned by  the peer (which may be None).

The RSMIterator is an async iterator. It operates like this:

* On the first call, it invokes the coroutine. It stores the iterable internally, as well as the RSM XSO which was returned. It yields the first item of the iterable.
* On subsequent calls, it yields the next item of the iterable until it is exhausted.
* When the iterable is exhausted, it invokes the coroutine again. The RSM XSO passed to the coroutine is constructed (using the internal RSM state as returned by the previous call) so that the next page is returned. Otherwise, this operates exactly like the first call. If the internal RSM XSO is None, or the coroutine returns an empty iterable, the iteration ends.

We can re-use this concept in DiscoClient and other places.

**Comment by sebastianriese** [comment]
Yup, this sounds about right (and generic in a nice way). I would perhaps also add an asnyc iterator that returns blocks as they come (some operations may be much easier/faster to do on a sequence of lists, instead of during element-wise iteration).

**Comment by horazont** [comment]
Good idea. Maybe the base implementation should just yield the blocks, and then a flattening adapter can be used if needed.

**Comment by sebastianriese** [comment]
Python seems to need aitertools.
